File: BinaryTree/php/breadthFirstValues.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/TreeNode.php';

function breadthFirstValues(?TreeNode $root): array
{
    if ($root === null) {
        return [];
    }

    $queue = [$root];
    $values = [];

    while ($queue) {
        /** @var TreeNode $current */
        $current = array_shift($queue);
        $values[] = $current->val;
        if ($current->left !== null) {
            $queue[] = $current->left;
        }
        if ($current->right !== null) {
            $queue[] = $current->right;
        }
    }

    return $values;
}

File: BinaryTree/go/breadthFirstValues.go
package binarytree

// BreadthFirstValues returns the node values level by level from left to right.
func BreadthFirstValues(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	values := make([]int, 0, 16)
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		values = append(values, node.Val)
		if node.Left != nil {
			queue = append(queue, node.Left)
		}
		if node.Right != nil {
			queue = append(queue, node.Right)
		}
	}

	return values
}
